src/sound/asio/sound.cpp

```diff
@@ -1218,3 +1235,75 @@ int64_t CSound::Flip64Bits ( const int64_t iIn )
 
     return iOut;
 }
+
+// Windows Native MIDI support
+void CSound::MidiStart()
+{
+    /* Get the number of MIDI In devices in this computer */
+    iMidiDevs = midiInGetNumDevs();
+    if ( iMidiDevs > MAX_MIDI_DEVS )
+    {
+        iMidiDevs = MAX_MIDI_DEVS;
+    }
+
+    // printf("Found %d MIDI device%s\n", iMidiDevs, iMidiDevs == 1 ? "" : "s");
+
+    // open all connected MIDI devices and set the callback function to handle incoming messages
+    for ( int i = 0; i < iMidiDevs; i++ )
+    {
+        MMRESULT result = midiInOpen ( &hMidiIn[i], i, (DWORD_PTR) MidiCallback, 0, CALLBACK_FUNCTION );
+
+        if ( result != MMSYSERR_NOERROR )
+        {
+            qWarning() << "! Failed to open MIDI input device. Error code: " << result;
+            hMidiIn[i] = 0;
+            return;
+        }
+
+        result = midiInStart ( hMidiIn[i] );
+
+        if ( result != MMSYSERR_NOERROR )
+        {
+            qWarning() << "! Failed to start MIDI input. Error code: " << result;
+            midiInClose ( hMidiIn[i] );
+            hMidiIn[i] = 0;
+            return;
+        }
+    }
+}
+
+void CSound::MidiStop()
+{
+    // stop MIDI if running
+    for ( int i = 0; i < iMidiDevs; i++ )
+    {
+        if ( hMidiIn[i] != 0 )
+        {
+            midiInStop ( hMidiIn[i] );
+            midiInClose ( hMidiIn[i] );
+        }
+    }
+}
+
+void CALLBACK CSound::MidiCallback ( HMIDIIN hMidiIn, UINT wMsg, DWORD_PTR dwInstance, DWORD_PTR dwParam1, DWORD_PTR dwParam2 )
+{
+    Q_UNUSED ( hMidiIn );
+    Q_UNUSED ( dwInstance );
+    Q_UNUSED ( dwParam2 );
+
+    if ( wMsg == MIM_DATA )
+    {
+        BYTE status = dwParam1 & 0xFF;
```

Comment:
I'd like to see a comment here what the mask does and why it works. Maybe a link to midi specifications or similar.

Reply:
Comments added, see be04d89